### PR TITLE
[performance](load) use vector instead of skiplist when insert agg keys

### PR DESCRIPTION
**Pull request by cnissnzg**
# Proposed changes

Issue Number: #18917 

## Problem summary

In the current Memtable implementation, we place the inserted data into the input_block and use additional data structures to maintain its orderliness: sometimes we aggregate data in input_block to reduce memory usage, sometimes we flush data out. In these cases, the incremental sorting approach is sufficient, with better performance than several single inserts in a skiplist.

In this PR, we:

1. Use incremental sorting in memtable to maintain orderliness, instead of skiplist.
2. Splitting the original collect function into two parts, sorting and aggregating, to improve readability and reduce unnecessary aggregation steps

Test result of stream load into unique key table with 3 replicas in 3 X 16c 64GB machines from @kaijchen. 

| test suite | before this change | after this change |
| --- | --- | --- |
| 7.5GB, 1 bucket, single_replica_read = false | 160s | 130s |
| 96GB, 960 bucket, single_replica_read = false | 6189s | 1360s |
| 96GB, 960 bucket, single_replica_read = true | 2318s | 919s |

## Checklist(Required)

* [x] Does it affect the original behavior
* [x] Has unit tests been added
* [x] Has document been added or modified
* [x] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/memtable.cpp [hunk 1]
```diff
@@ -286,32 +292,90 @@ void MemTable::_collect_vskiplist_results() {
...
+        vectorized::MutableBlock mutable_block =
+                vectorized::MutableBlock::build_mutable_block(&in_block);
+        _vec_row_comparator->set_block(&mutable_block);
+        auto new_row_it = std::next(_row_in_blocks.begin(), _last_sorted_pos);
+        // sort new rows
+        std::sort(new_row_it, _row_in_blocks.end(),
+                  [this](const RowInBlock* l, const RowInBlock* r) -> bool {
+                      auto value = (*(this->_vec_row_comparator))(l, r);
+                      if (value == 0) {
+                          return l->_row_pos > r->_row_pos;
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:305]: add a stats here, e.g. same_keys_num++;

File: be/src/olap/memtable.cpp [hunk 2]
```diff
@@ -286,32 +292,90 @@ void MemTable::_collect_vskiplist_results() {
...
+                      } else {
+                          return value < 0;
+                      }
+                  });
+        // merge new rows and old rows
+        std::inplace_merge(_row_in_blocks.begin(), new_row_it, _row_in_blocks.end());
+        _last_sorted_pos = _row_in_blocks.size();
+        std::vector<RowInBlock*> temp_row_in_blocks(_last_sorted_pos);
+        std::vector<size_t> need_merge_idxs; // use bitset?
+        bool need_init_agg = true;
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:315]: if (same_keys_num == 0) skip agg.

File: be/src/olap/memtable.cpp [hunk 3]
```diff
@@ -21,10 +21,13 @@
 #include <gen_cpp/olap_file.pb.h>
 
 #include <algorithm>
+#include <bitset>
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:24]: bitset is not used.

File: be/src/olap/memtable.cpp [hunk 4]
```diff
@@ -267,79 +218,137 @@ void MemTable::_aggregate_two_row_in_block(RowInBlock* new_row, RowInBlock* row_
...
+    std::vector<int> row_pos_vec;
+    DCHECK(in_block.rows() <= std::numeric_limits<int>::max());
+    row_pos_vec.reserve(in_block.rows());
+    for (int i = 0; i < _row_in_blocks.size(); i++) {
+        row_pos_vec.emplace_back(_row_in_blocks[i]->_row_pos);
+    }
+    _output_mutable_block.add_rows(&in_block, row_pos_vec.data(),
+                                   row_pos_vec.data() + in_block.rows());
+}
+int MemTable::_sort_row_in_blocks() {
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:231]: Name _sort is enough.

File: be/src/olap/memtable.cpp [hunk 5]
```diff
@@ -267,79 +218,137 @@ void MemTable::_aggregate_two_row_in_block(RowInBlock* new_row, RowInBlock* row_
                                  new_row->_row_pos, nullptr);
     }
 }
+void MemTable::_add_rows_from_block(vectorized::Block& in_block) {
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:221]: prepare_block_for_flush is a better name

File: be/src/olap/memtable.cpp [hunk 6]
```diff
@@ -267,79 +218,137 @@ void MemTable::_aggregate_two_row_in_block(RowInBlock* new_row, RowInBlock* row_
...
+            (*_vec_row_comparator)(prev_row, _row_in_blocks[i]) == 0) {
+            if (need_init_agg) {
+                prev_row->init_agg_places(
+                        _arena->aligned_alloc(_total_size_of_aggregate_states, 16),
+                        _offsets_of_aggregate_states.data());
+                for (auto cid = _schema->num_key_columns(); cid < _schema->num_columns(); cid++) {
+                    auto col_ptr = _input_mutable_block.mutable_columns()[cid].get();
+                    auto data = prev_row->agg_places(cid);
+                    _agg_functions[cid]->create(data);
+                    _agg_functions[cid]->add(
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:288]: _agg_functions may be inited in last call of merge_row_in_block. we should push init_agg as a member of MemTable.

Comment (dataroaring) [reply be/src/olap/memtable.cpp:288]: And we put init code in a function called init_agg_functions()

File: be/src/olap/memtable.cpp [hunk 7]
```diff
@@ -267,79 +218,137 @@ void MemTable::_aggregate_two_row_in_block(RowInBlock* new_row, RowInBlock* row_
...
+                           }
+                       });
+    _last_sorted_pos = _row_in_blocks.size();
+    return same_keys_num;
+}
+
 template <bool is_final>
-void MemTable::_collect_vskiplist_results() {
-    VecTable::Iterator it(_vec_skip_list.get());
+void MemTable::_merge_row_in_blocks() {
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:265]: aggregate is a better name.

File: be/src/olap/memtable.cpp [hunk 8]
```diff
@@ -267,79 +218,137 @@ void MemTable::_aggregate_two_row_in_block(RowInBlock* new_row, RowInBlock* row_
...
+                need_init_agg = false;
             }
+            _merged_rows++;
+            _aggregate_two_row_in_block(_row_in_blocks[i], prev_row);
+        } else {
+            prev_row = _row_in_blocks[i];
+            temp_row_in_blocks.push_back(prev_row);
+            need_init_agg = true;
+        }
+    }
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:302]: we should stat merged rows in one run and we can optimize move operations below.

**Review by dataroaring — CHANGES_REQUESTED** [review]

File: be/src/olap/memtable.cpp [hunk 9]
```diff
@@ -267,102 +230,152 @@ void MemTable::_aggregate_two_row_in_block(RowInBlock* new_row, RowInBlock* row_
...
-        _vec_skip_list.reset();
+void MemTable::prepare_block_for_flush(vectorized::Block& in_block) {
+    std::vector<int> row_pos_vec;
+    DCHECK(in_block.rows() <= std::numeric_limits<int>::max());
+    row_pos_vec.reserve(in_block.rows());
+    for (int i = 0; i < _row_in_blocks.size(); i++) {
+        row_pos_vec.emplace_back(_row_in_blocks[i]->_row_pos);
     }
+    _output_mutable_block.add_rows(&in_block, row_pos_vec.data(),
+                                   row_pos_vec.data() + in_block.rows());
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:241]: we should skip memcopy for dupliate without keys like _collect_dup_table_without_keys does.

File: be/src/olap/memtable.cpp [hunk 10]
```diff
@@ -267,102 +230,152 @@ void MemTable::_aggregate_two_row_in_block(RowInBlock* new_row, RowInBlock* row_
...
+                               return value < 0;
+                           }
+                       });
+    _last_sorted_pos = _row_in_blocks.size();
+    return same_keys_num;
+}
+
+template <bool is_final>
+void MemTable::_aggregate_one_row(RowInBlock* row,
+                                  const vectorized::ColumnsWithTypeAndName& block_data,
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:278]: finalize_one_row is a better name.

File: be/src/olap/memtable.cpp [hunk 11]
```diff
@@ -144,16 +144,14 @@ void MemTable::_init_agg_functions(const vectorized::Block* block) {
 }
 
 MemTable::~MemTable() {
-    if (_vec_skip_list != nullptr && _keys_type != KeysType::DUP_KEYS) {
-        VecTable::Iterator it(_vec_skip_list.get());
-        for (it.SeekToFirst(); it.Valid(); it.Next()) {
+    if (_keys_type != KeysType::DUP_KEYS) {
+        for (auto it = _row_in_blocks.begin(); it != _row_in_blocks.end(); it++) {
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:148]: if (!it->has_init_agg()) {
    continue;
}

**Comment by dataroaring** [comment]
run buildall

**Review by dataroaring — CHANGES_REQUESTED** [review]

**Comment by dataroaring** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

File: be/src/olap/memtable.cpp [hunk 12]
```diff
@@ -440,7 +463,13 @@ Status MemTable::flush() {
 Status MemTable::_do_flush(int64_t& duration_ns) {
     SCOPED_CONSUME_MEM_TRACKER(_flush_mem_tracker);
     SCOPED_RAW_TIMER(&duration_ns);
-    _collect_vskiplist_results<true>();
+    _sort();
+    if (_keys_type == KeysType::DUP_KEYS) {
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:467]: int same_keys_num = _sort();
if (_keys_type == KeysType::DUP_KEYS || same_keys_num == 0)

**Review by dataroaring — CHANGES_REQUESTED** [review]

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 36.15 seconds
 stream load tsv:          433 seconds loaded 74807831229 Bytes, about 164 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          78.4 seconds inserted 10000000 Rows, about 127K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230520125426_clickbench_pr_147718.html

**Comment by dataroaring** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 36.22 seconds
 stream load tsv:          430 seconds loaded 74807831229 Bytes, about 165 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          61 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          78.0 seconds inserted 10000000 Rows, about 128K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230521092756_clickbench_pr_147843.html

**Comment by cnissnzg** [comment]
run buildall

**Comment by cnissnzg** [comment]
run buildall

**Comment by cnissnzg** [comment]
run buildall

**Comment by kaijchen** [comment]
Test result of stream load into unique key table with 3 replicas.

| test suite | before this change | after this change |
| --- | --- | --- |
| 7.5GB, 1 bucket, single_replica_read = false | 160s | 130s |
| 96GB, 960 bucket, single_replica_read = false | 6189s | 1360s |
| 96GB, 960 bucket, single_replica_read = true | 2318s | 919s |

**Review by dataroaring — DISMISSED** [review]
LGTM

File: be/src/olap/memtable.h [hunk 13]
```diff
@@ -178,12 +186,18 @@ class MemTable {
     //for vectorized
     vectorized::MutableBlock _input_mutable_block;
     vectorized::MutableBlock _output_mutable_block;
+    size_t _last_sorted_pos = 0;
 
+    //return number of same keys
+    int _sort();
+    template <bool is_final>
+    void finalize_one_row(RowInBlock* row, const vectorized::ColumnsWithTypeAndName& block_data,
```
Comment (kaijchen) [inline be/src/olap/memtable.h:194]: Private method name should starts with an underscore
 
```suggestion
    void _finalize_one_row(RowInBlock* row, const vectorized::ColumnsWithTypeAndName& block_data,
```

File: be/src/olap/memtable.h [hunk 14]
```diff
@@ -178,12 +186,18 @@ class MemTable {
...
+                          int row_pos);
     template <bool is_final>
-    void _collect_vskiplist_results();
-    void _collect_dup_table_with_keys();
-    void _collect_dup_table_without_keys();
+    void _aggregate();
+    void prepare_block_for_flush(vectorized::Block& in_block);
```
Comment (kaijchen) [inline be/src/olap/memtable.h:198]:
```suggestion
    void _prepare_block_for_flush(vectorized::Block& in_block);
```

File: be/src/olap/memtable.h [hunk 15]
```diff
@@ -178,12 +186,18 @@ class MemTable {
...
     bool _is_first_insertion;
+    std::function<void(int64_t)> delta_writer_callback;
```
Comment (kaijchen) [inline be/src/olap/memtable.h:200]:
```suggestion
    std::function<void(int64_t)> _delta_writer_callback;
```

File: be/src/olap/memtable.h [hunk 16]
```diff
@@ -113,8 +123,8 @@ class MemTable {
```
Comment (kaijchen) [inline be/src/olap/memtable.h:123]: This can be removed.

```cpp
private:
    using VecTable = SkipList<RowInBlock*, RowInBlockComparator>;
```

**Review by kaijchen — COMMENTED** [review]
Thanks @cnissnzg for the improvement. Mostly LGTM, just some minor comments about code style.

**Comment by dataroaring** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.55 seconds
 stream load tsv:          421 seconds loaded 74807831229 Bytes, about 169 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          75.3 seconds inserted 10000000 Rows, about 132K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523014327_clickbench_pr_148572.html

File: be/src/olap/memtable.cpp [hunk 17]
```diff
@@ -425,6 +443,7 @@ Status MemTable::flush() {
     // and use the ids to load segment data file for calc delete bitmap.
     int64_t atomic_num_segments_before_flush = _rowset_writer->get_atomic_num_segment();
     RETURN_IF_ERROR(_do_flush(duration_ns));
+    _delta_writer_callback(_merged_rows);
```
Comment (kaijchen) [inline be/src/olap/memtable.cpp:446]: Why do we need this callback?

AFAIK, `_merged_rows` is already increased in 

```cpp
Status DeltaWriter::_flush_memtable_async() {
    _merged_rows += _mem_table->merged_rows();
    return _flush_token->submit(std::move(_mem_table));
}
```

The callstack (one of the code paths):

```cpp
DeltaWriter::_flush_memtable_async();
=> FlushToken::submit();
=> FlushToken::_flush_memtable();
=> new MemtableFlushTask();
=> MemtableFlushTask::run();
=> MemTable::flush(); // in the latest master, equivalent to MemTable::need_flush()
```

Comment (kaijchen) [reply be/src/olap/memtable.cpp:446]: OK I got it, `_mem_table->merged_rows()` was updated in `MemTable::_do_flush()`.

```cpp
    RETURN_IF_ERROR(_do_flush(duration_ns));
    delta_writer_callback(_merged_rows);
```

The question is, will it be added twice?

```cpp
// in DeltaWriter::_flush_memtable_async()
    _merged_rows += _mem_table->merged_rows(); // what if merged_rows() is non-zero here?

// in MemTable::flush() or MemTable::need_flush()
    _delta_writer_callback(_merged_rows)
```

I wonder if there is a better way to make this more clear to read.

Comment (cnissnzg) [reply be/src/olap/memtable.cpp:446]: Oh yes, I think we should remove this line in _flush_memtable_async.
Now we only use _merged_rows to check correctness, and we can't get correct _merged_rows until flush has finished.

**Comment by cnissnzg** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 32.98 seconds
 stream load tsv:          429 seconds loaded 74807831229 Bytes, about 166 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          76.1 seconds inserted 10000000 Rows, about 131K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523075531_clickbench_pr_148803.html

**Review by kaijchen — APPROVED** [review]
LGTM, thanks @cnissnzg for the improvement.
I'll post detailed result of performance test later.

**Comment by cnissnzg** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 32.96 seconds
 stream load tsv:          429 seconds loaded 74807831229 Bytes, about 166 MB/s
 stream load json:         23 seconds loaded 2358488459 Bytes, about 97 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          76.2 seconds inserted 10000000 Rows, about 131K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523104117_clickbench_pr_148933.html

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**